File: src/main/java/com/chatchatabc/jpademojava/application/rest/PassportController.java
package com.chatchatabc.jpademojava.application.rest;

import com.chatchatabc.jpademojava.application.dto.ErrorContent;
import com.chatchatabc.jpademojava.application.dto.passport.*;
import com.chatchatabc.jpademojava.domain.model.Passport;
import com.chatchatabc.jpademojava.domain.model.User;
import com.chatchatabc.jpademojava.domain.repository.PassportRepository;
import com.chatchatabc.jpademojava.domain.repository.UserRepository;
import com.chatchatabc.jpademojava.domain.service.PassportService;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Optional;

@RestController
@RequestMapping("/api/passport")
public class PassportController {
    @Autowired
    private PassportService passportService;
    @Autowired
    private PassportRepository passportRepository;
    @Autowired
    private UserRepository userRepository;

    private final ModelMapper modelMapper = new ModelMapper();

    /**
     * Get all passports
     *
     * @param pageable
     * @return
     */
    @GetMapping("/get")
    public ResponseEntity<Page<Passport>> getPassports(
            Pageable pageable
    ) {
        try {
            Page<Passport> passports = passportRepository.findAll(pageable);
            return ResponseEntity.ok(passports);
        } catch (Exception e) {
            return ResponseEntity.badRequest().build();
        }
    }

    /**
     * Get passport by user
     *
     * @param userId
     * @return
     */
    @GetMapping("/get/{userId}")
    public ResponseEntity<Passport> getPassportByUser(
            @PathVariable String userId
    ) {
        try {
            Optional<User> user = userRepository.findById(userId);
            if (user.isEmpty()) {
                throw new Exception("User not found");
            }
            Optional<Passport> passport = passportRepository.findPassportByUser(user.get());
            if (passport.isEmpty()) {
                throw new Exception("Passport not found");
            }
            return ResponseEntity.ok(passport.get());
        } catch (Exception e) {
            return ResponseEntity.badRequest().build();
        }
    }

    /**
     * Create a passport for a user
     *
     * @param request
     * @param userId
     * @return
     */
    @PostMapping("/create/{userId}")
    public ResponseEntity<PassportResponse> createPassport(
            @RequestBody PassportCreateRequest request,
            @PathVariable String userId
    ) {
        try {
            Passport passport = modelMapper.map(request, Passport.class);
            User updatedUser = passportService.create(userId, passport);
            return ResponseEntity.ok(new PassportResponse(updatedUser, null));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(new PassportResponse(null, new ErrorContent("Update Passport Error", e.getMessage())));
        }
    }

    /**
     * Update a passport
     *
     * @param passportId
     * @param request
     * @return
     */
    @PutMapping("/update/{passportId}")
    public ResponseEntity<PassportUpdateResponse> updatePassport(
            @PathVariable String passportId,
            @RequestBody PassportUpdateRequest request
    ) {
        try {
            Passport newPassportInfo = modelMapper.map(request, Passport.class);
            Passport updatedPassport = passportService.update(passportId, newPassportInfo);
            return ResponseEntity.ok(new PassportUpdateResponse(updatedPassport, null));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(
                    new PassportUpdateResponse(null, new ErrorContent("Update Passport Error", e.getMessage()))
            );
        }
    }

    /**
     * Delete a passport
     *
     * @param userId
     * @return
     */
    @DeleteMapping("/delete/{userId}")
    public ResponseEntity<PassportDeleteResponse> deletePassport(
            @PathVariable String userId
    ) {
        try {
            User updatedUser = passportService.delete(userId);
            return ResponseEntity.ok(new PassportDeleteResponse(updatedUser, null));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(
                    new PassportDeleteResponse(null, new ErrorContent("Delete Passport Error", e.getMessage()))
            );
        }
    }
}

File: src/main/kotlin/com/chatchatabc/jpademo/application/rest/PassportController.kt
package com.chatchatabc.jpademo.application.rest

import com.chatchatabc.jpademo.application.dto.ErrorContent
import com.chatchatabc.jpademo.application.dto.passport.*
import com.chatchatabc.jpademo.domain.model.Passport
import com.chatchatabc.jpademo.domain.repository.PassportRepository
import com.chatchatabc.jpademo.domain.repository.UserRepository
import com.chatchatabc.jpademo.domain.service.PassportService
import org.modelmapper.ModelMapper
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/passport")
class PassportController(
    private val passportService: PassportService,
    private val passportRepository: PassportRepository,
    private val userRepository: UserRepository
) {
    private val mapper = ModelMapper()

    /**
     * Get Passports
     */
    @GetMapping("/get")
    fun getPassports(
        pageable: Pageable
    ): ResponseEntity<Page<Passport>> {
        return try {
            val passports = passportRepository.findAll(pageable)
            ResponseEntity.ok(passports)
        } catch (e: Exception) {
            ResponseEntity.badRequest().build()
        }
    }

    /**
     * Get Passport by User
     */
    @GetMapping("/get/{userId}")
    fun getPassport(
        @PathVariable userId: String
    ): ResponseEntity<Passport> {
        return try {
            val user = userRepository.findById(userId)
            val passport = passportRepository.findPassportByUser(user.get())
            ResponseEntity.ok(passport.get())
        } catch (e: Exception) {
            ResponseEntity.badRequest().build()
        }
    }

    /**
     * Create Passport
     */
    @PostMapping("/create/{userId}")
    fun createPassport(
        @RequestBody request: PassportCreateRequest,
        @PathVariable userId: String
    ): ResponseEntity<PassportResponse> {
        return try {
            val passport = mapper.map(request, Passport::class.java)
            val updatedUser = passportService.create(userId, passport)
            ResponseEntity.ok(PassportResponse(updatedUser, null))
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
                PassportResponse(
                    null, ErrorContent(
                        "Create Passport Error", e.message ?: "Unknown Error"
                    )
                )
            )
        }
    }

    /**
     * Update Passport
     */
    @PutMapping("/update/{passportId}")
    fun updatePassport(
        @PathVariable passportId: String,
        @RequestBody request: PassportUpdateRequest
    ): ResponseEntity<PassportUpdateResponse> {
        return try {
            val newPassportInfo = mapper.map(request, Passport::class.java)
            val updatedPassport = passportService.update(passportId, newPassportInfo)
            ResponseEntity.ok(PassportUpdateResponse(updatedPassport, null))
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(
                PassportUpdateResponse(
                    null, ErrorContent(
                        "Update Passport Error", e.message ?: "Unknown Error"
                    )
                )
            )
        }
    }

    /**
     * Delete Passport
     */
    @DeleteMapping("/delete/{userId}")
    fun deletePassport(
        @PathVariable userId: String
    ): ResponseEntity<PassportDeleteResponse> {
        return try {
            val user = passportService.delete(userId)
            ResponseEntity.ok(PassportDeleteResponse(user, null))
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(
                PassportDeleteResponse(
                    null, ErrorContent(
                        "Delete Passport Error", e.message ?: "Unknown Error"
                    )
                )
            )
        }
    }
}